crates/rpc/src/method/get_storage_proof.rs

```diff
@@ -0,0 +1,784 @@
+    }
+}
+
+// Doing this manually since `generate_rpc_error_subset!`
+// does not support enum struct variants.
+impl From<Error> for crate::error::ApplicationError {
+    fn from(e: Error) -> Self {
+        match e {
+            Error::ProofLimitExceeded { limit, requested } => {
+                Self::ProofLimitExceeded { limit, requested }
+            }
+            Error::BlockNotFound => Self::BlockNotFound,
+            Error::Internal(internal) => Self::Internal(internal),
+            Error::StorageProofNotSupported => Self::StorageProofNotSupported,
+            Error::ProofMissing => Self::ProofMissing,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Input {
+    pub block_id: BlockId,
+    pub class_hashes: Option<Vec<ClassHash>>,
+    pub contract_addresses: Option<Vec<ContractAddress>>,
+    pub contracts_storage_keys: Option<Vec<ContractStorageKeys>>,
+}
+
+impl DeserializeForVersion for Input {
+    fn deserialize(value: crate::dto::Value) -> Result<Self, serde_json::Error> {
+        value.deserialize_map(|value| {
+            Ok(Self {
+                block_id: value.deserialize("block_id")?,
+                class_hashes: value.deserialize_optional_array("class_hashes", |value| {
+                    value.deserialize().map(ClassHash)
+                })?,
+                contract_addresses: value
+                    .deserialize_optional_array("contract_addresses", |value| {
+                        value.deserialize().map(ContractAddress)
+                    })?,
+                contracts_storage_keys: value
+                    .deserialize_optional_array("contracts_storage_keys", |value| {
+                        value.deserialize()
+                    })?,
+            })
+        })
+    }
+}
+
+/// Wrapper around [`TrieNode`] to implement [`SerializeForVersion`].
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct ProofNode(TrieNode);
+
+impl SerializeForVersion for ProofNode {
+    fn serialize(
+        &self,
+        serializer: crate::dto::serialize::Serializer,
+    ) -> Result<crate::dto::serialize::Ok, crate::dto::serialize::Error> {
+        let mut s = serializer.serialize_struct()?;
+        match &self.0 {
+            TrieNode::Binary { left, right } => {
+                s.serialize_field("left", &left)?;
+                s.serialize_field("right", &right)?;
+            }
+            TrieNode::Edge { child, path } => {
+                let p = Felt::from_bits(path).unwrap();
+                let len = path.len();
+
+                s.serialize_field("path", &p)?;
+                s.serialize_field("length", &len)?;
+                s.serialize_field("child", &child)?;
+            }
+        }
+        s.end()
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct NodeHashToNodeMapping {
+    node_hash: Felt,
+    node: ProofNode,
+}
+
+impl From<TrieNode> for NodeHashToNodeMapping {
+    fn from(node: TrieNode) -> Self {
+        let node_hash = node.hash::<PedersenHash>();
```

Comment:
A possible optimization would be using the node hash from storage instead of re-computing it here for each node.